### ISSUE TITLE
[no-unused-vars] False positives with members of exported namespace in .d.ts file

### ISSUE THREAD

**Issue by andersk** [typescript-eslint/typescript-eslint#2867]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint"],
  "rules": {
    "@typescript-eslint/no-unused-vars": "error"
  }
}
```

```TS
// foo.d.ts
export namespace Foo {
  const bar: string;
  namespace Baz {
    const quux: string;
  }
}
```

**Expected Result**

No error. In a `.d.ts` file, members of an exported namespace are implicitly exported, unless the namespace contains an explicit braced `export {}` declaration.

**Actual Result**

```
/tmp/test/foo.d.ts
  2:9   error  'bar' is defined but never used   @typescript-eslint/no-unused-vars
  3:13  error  'Baz' is defined but never used   @typescript-eslint/no-unused-vars
  4:11  error  'quux' is defined but never used  @typescript-eslint/no-unused-vars

✖ 3 problems (3 errors, 0 warnings)
```

**Additional Info**

Similar to #2456, which was about a different kind of implicit export in `.d.ts` files, namely interface declarations.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.9.1` |
| `@typescript-eslint/parser`        | `4.9.1` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.15.0` |
| `node`                             | `14.15.1` |

**Comment by bradzacher** [comment]
Why are you `export`ing your namespace here?

Also in declaration files - you should be `declare`ing your namespaces.

**Comment by andersk** [comment]
@bradzacher In a `.d.ts` file, `export declare` is the same as just `export`. It declares a namespace that’s an exported member of the module, to be used like `import {Foo} from "./foo"; Foo.Baz.quux`.

Just `declare namespace` without `export` would be different; that would declare a namespace that’s a global variable, to be used like `import "./foo"; Foo.Baz.quux`

**Comment by bradzacher** [comment]
How are you creating your declaration files for your modules?

I thought the general pattern was

```ts
declare module './foo' {
  export namespace Foo {
    // ..
  }
}
```

and not just declaring your namespaces in the root.

**Comment by andersk** [comment]
With a text editor?

A `foo.d.ts` file can either (a) declare types for third-party modules with `declare module 'lodash'`, or (b) declare types for `foo.js` in the same directory at the root. For a module in the same directory, `declare module './foo'` is actually illegal (`TS2436: Ambient module declaration cannot specify relative module name`); you can only do (b).

**Comment by bradzacher** [comment]
The issue is that our selectors target `Program > Namespace`, so because your namespace is `exported` it doesn't match this.

https://github.com/typescript-eslint/typescript-eslint/blob/84a77cc7e4e6bba2519e22c6d5772fc50961dedc/packages/eslint-plugin/src/rules/no-unused-vars.ts#L259

Happy to accept a PR - simple fix.

**Comment by bradzacher** [comment]
your config is wrong. `no-unused-vars` is the base rule.
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#i-am-using-a-rule-from-eslint-core-and-it-doesnt-work-correctly-with-typescript-code

Please read the docs before raising an issue.

**Comment by dopecodez** [comment]
I'll take a look at this 👨🏽‍💻

**Comment by tryggvigy** [comment]
Just upgraded our very large repo to
```diff
-    "@typescript-eslint/eslint-plugin": "3.9.1",
-    "@typescript-eslint/parser": "3.9.1",
+    "@typescript-eslint/eslint-plugin": "4.9.1",
+    "@typescript-eslint/parser": "4.9.1",
-    "typescript": "3.9.6",
+    "typescript": "4.1.3",
```

and seeing a lot of `@typescript-eslint/no-unused-vars` errors. They seem to be false positives that trigger either by optional arguments or optional chaining. I have a minimal case here:

![foo](https://user-images.githubusercontent.com/2373958/102084235-c4f74080-3e0c-11eb-80ca-d5032f7d9442.gif)

As you can see removing optional chaining fixes the issue.

```ts
function toURI(a: any): any {
  return a;
}

export const isSameContext = (
  oneContextUri?: string | null,
  anotherContextUri?: string | null,
): boolean => {
  if (!oneContextUri || !anotherContextUri) return false;

  const contextUri1 = toURI(oneContextUri);
  const contextUri2 = toURI(anotherContextUri);

  return contextUri1 && contextUri1.isSameIdentity(contextUri2);
};
```

**Comment by bradzacher** [comment]
@tryggvigy that issue is completely unrelated to this one.

**Comment by joaoeudes7** [comment]
I imagine this is not just happening in the .d.ts files.
```ts
export type GetHistoricBankUseCaseParams = {
    userId: string,
    bankId: string
}

export class GetHistoricBankUseCase implements IUseCase {
    constructor(
        private readonly historicRepository: HistoricRepository
    ) {}

    async execute(data: GetHistoricBankUseCaseParams) {
        const { bankId, userId } = data;
        const historic = await this.historicRepository.getHistoricBank(userId, bankId);
    
        return historic;
    }
}
```

I receive the error: `'historicRepository' is defined but never used.eslintno-unused-vars`

**Comment by bradzacher** [comment]
@joaoeudes7 Please don't comment on existing issues unless they are related to your issue. An issue about false positives in `.d.ts` files is obviously unrelated to false positives in function arguments.

But that aside. Please read the FAQ. You're using the wrong rule.

**Comment by joaoeudes7** [comment]
@bradzacher sorry to hesitate,  but I believe the problem is related, since what was happening was a false positovo, where a parameter of the constructor was considered unused even if it was used, after applying the rule:"@typescript-eslint/no-unused-vars": "error"` started to work as it should (reporting only the unused).

But sorry to disturb, I was just reporting that the same was happening in .ts files.

**Comment by danielrsantana** [comment]
> Same for me. Using version 4.8.2
> 
> ![image](https://user-images.githubusercontent.com/11296189/101991081-9a8d7200-3caa-11eb-9880-eb30cfac1693.png)
> 
> With version 4.9.1
> 
> ![image](https://user-images.githubusercontent.com/11296189/101991107-c27cd580-3caa-11eb-9adc-b60648fdf0d6.png)
> 
> where the issues presented are related to interfaces and I don't even use Namespace.
> 
> ![image](https://user-images.githubusercontent.com/11296189/101991119-e17b6780-3caa-11eb-91ff-de025db8005c.png)
> 
> ![image](https://user-images.githubusercontent.com/11296189/101991137-fd7f0900-3caa-11eb-9371-6b411f97fd06.png)

How can this be marked as of topic? I’m unable to update version because latest present false positive on interfaces.

Please take a look.
Tks

**Comment by danielrsantana** [comment]
> @bradzacher sorry to hesitate, but I believe the problem is related, since what was happening was a false positive, where a parameter of the constructor was considered unused even if it was used, after applying the rule:"@typescript-eslint/no-unused-vars": "error"` started to work as it should (reporting only the unused).
> 
> But sorry to disturb, I was just reporting that the same was happening in .ts files.

This is the expected behavior , @joaoeudes7. Your constructor is not doing anything with the repository variable.

Cheers all

**Comment by bradzacher** [comment]
@danielrsantana you ignored my response.
https://github.com/typescript-eslint/typescript-eslint/issues/2867#issuecomment-743798470

**Comment by bradzacher** [comment]
To anyone seeing this issue.

This issue is ***ONLY*** about `@typescript-eslint/no-unused-vars` reporting false-positives for exported namespaces in `.d.ts` files.

***ALL OTHER ISSUES ARE NOT RELATED TO THIS ISSUE.***

# [We have an FAQ you should read before commenting](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#i-am-using-a-rule-from-eslint-core-and-it-doesnt-work-correctly-with-typescript-code)

**Comment by smolster** [comment]
Any update here? I'm seeing this issue in version 4.23.0 of the plugin.

**Comment by bradzacher** [comment]
@JounQin I'm unable to repro that against master.

![image](https://user-images.githubusercontent.com/7462525/118547263-f5217300-b70d-11eb-9537-f355176a98be.png)

If I were to hazard a guess based on the first false-positive you listed; you've got an old version of the package installed and running.

**Comment by TheDutchCoder** [comment]
This also happens for the vue shim:
```ts
/* eslint-disable @typescript-eslint/no-explicit-any */
declare module '*.vue' {
  import type { DefineComponent } from 'vue'
  const component: DefineComponent<Record<string, unknown>, Record<string, unknown>, any>
  export default component
}
```

**Comment by andersk** [comment]
@TheDutchCoder Your code has no exported `namespace`s, so your issue is not this issue. (You haven’t indicated what your issue is, since at least the current release of `@typescript-eslint` doesn’t give any errors for your code, but in any event you’ll need to ask about it somewhere else.)

**Comment by jakebailey** [comment]
I was hoping to add this rule to DefinitelyTyped (https://github.com/microsoft/DefinitelyTyped-tools/pull/968), but noticed this problem as well as most namespaces on DT have implicit export modifiers. I may look into this (though please don't take this as [cookie licking](https://devblogs.microsoft.com/oldnewthing/20091201-00/?p=15843) and please still attempt it even if I forget or fail).

A gotcha is that this namespace does _not_ have any unused vars:

```ts
export namespace Foo {
    const value: 1234;

    export import Something = something.Something;
}
```

Yay namespaces.
